File: backend/app/schemas/dependant_query.py
"""Dependant selection — the roster's other population.

Dependants are not employees, so they need their own predicates (relationship,
link state, the status of a portal self-add). But a dependant's *cohort* is its
sponsoring employee's cohort, so the employee half is the existing
``MemberFilters`` nested verbatim rather than a parallel vocabulary — that is
what makes "show me the dependants of everyone in Subsidiary B on Plan 3" one
query instead of two screens.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.member_query import AgeFilter, FacetValue, MemberFilters

# Mirrors ``models.dependant`` — a portal self-add lands as ``pending_approval``
# and only becomes coverage-bearing once a broker approves it.
DependantStatusStr = Literal["active", "pending_approval", "rejected", "terminated"]
# ``classify_relationship`` answers spouse/child/None; "other" is that None,
# named. It is the bucket parents and unclassifiable roster wording fall into,
# and is worth filtering on precisely because it is where data problems collect.
DependantRoleStr = Literal["spouse", "child", "other"]
LinkStateStr = Literal["any", "linked", "unlinked"]


class DependantFilters(BaseModel):
    q: str | None = Field(default=None, max_length=255)
    # EMPTY means the default view: active only. There is deliberately no "all"
    # token — the UI ticks every box, so what was asked for is always explicit.
    statuses: list[DependantStatusStr] = Field(default_factory=list, max_length=4)
    # Raw roster wording ("Spouse", "Son", "Daughter"). Free text with no enum,
    # so the vocabulary is served from the facets, never hardcoded.
    relationships: list[str] = Field(default_factory=list, max_length=200)
    roles: list[DependantRoleStr] = Field(default_factory=list, max_length=3)
    link_state: LinkStateStr = "any"
    link_methods: list[str] = Field(default_factory=list, max_length=20)
    # Age-Next-Birthday at the benefit year's start — the same convention every
    # dependant eligibility window uses, so "children ageing out" agrees with
    # the rules that will actually drop them.
    age: AgeFilter | None = None
    # The SPONSORING employee. Nested rather than duplicated, so a dependant can
    # be filtered by category, product, plan, entity or any roster attribute.
    # Setting it necessarily excludes unlinked dependants — they have no
    # employee to test.
    employee: MemberFilters | None = None

    def has_filters(self) -> bool:
        return bool(
            self.q
            or self.statuses
            or self.relationships
            or self.roles
            or self.link_methods
            or self.age
            or self.link_state != "any"
            or (self.employee is not None and self.employee.has_filters())
        )


class DependantQueryListIn(BaseModel):
    query: DependantFilters = Field(default_factory=DependantFilters)
    offset: int = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=200)


# ── Facets ──────────────────────────────────────────────────────────────────


class DependantFacetsOut(BaseModel):
    """Vocabulary + headcounts for the Dependants filter bar.

    Every facet except ``statuses`` counts the DEFAULT view (active dependants),
    because that is the population those filters narrow. ``statuses`` spans every
    dependant in the year, since it is the control that *widens* the population —
    the same split ``MemberFacetsOut`` makes with ``terminated_total``.
    """

    active_total: int
    all_statuses_total: int
    linked: int
    unlinked: int
    statuses: list[FacetValue]
    relationships: list[FacetValue]
    roles: list[FacetValue]
    link_methods: list[FacetValue]

File: frontend/src/api/dependantQuery.ts
/**
 * Dependant selection — the Dependants tab's filter vocabulary and its rows.
 *
 * The employee half of the query is a nested `MemberQuery`, not a parallel
 * vocabulary, so filtering dependants by cohort/product/plan/entity resolves
 * through exactly the chain the Employees tab uses.
 */
import { useQuery } from "@tanstack/react-query";
import { api } from "@/api/client";
import { useSession } from "@/stores/session";
import type { FacetValue, MemberQuery } from "@/api/memberQuery";
import type { DependantList } from "@/types";

export type DependantStatus =
  | "active"
  | "pending_approval"
  | "rejected"
  | "terminated";
export type DependantRole = "spouse" | "child" | "other";
export type LinkState = "any" | "linked" | "unlinked";

export interface DependantQuery {
  q?: string | null;
  /** EMPTY means the default view: active only. There is no "all" token — the
   *  UI ticks every box, so what was asked for is always explicit. */
  statuses?: DependantStatus[];
  /** Raw roster wording ("Spouse", "Son"). Served from the facets, never
   *  hardcoded — the vocabulary is free text with no enum. */
  relationships?: string[];
  roles?: DependantRole[];
  link_state?: LinkState;
  link_methods?: string[];
  age?: { min?: number | null; max?: number | null } | null;
  /** The SPONSORING employee. Setting it necessarily excludes unlinked
   *  dependants — there is no employee to test them against. */
  employee?: MemberQuery | null;
}

export interface DependantFacets {
  active_total: number;
  all_statuses_total: number;
  linked: number;
  unlinked: number;
  /** Spans every status — it is the control that WIDENS the population. */
  statuses: FacetValue[];
  relationships: FacetValue[];
  roles: FacetValue[];
  link_methods: FacetValue[];
}

export function useDependantFacets(policyYearId: string | undefined) {
  const clientId = useSession((s) => s.activeClientId);
  return useQuery({
    queryKey: ["dependants", "facets", clientId, policyYearId],
    queryFn: () =>
      api.get<DependantFacets>(`/policy-years/${policyYearId}/dependant-facets`),
    enabled: !!policyYearId,
  });
}

export function useDependantQueryList(
  policyYearId: string | undefined,
  query: DependantQuery,
  page: { offset: number; limit: number },
) {
  const clientId = useSession((s) => s.activeClientId);
  return useQuery({
    // Prefixed "dependants" — see the note on the member listing key.
    queryKey: [
      "dependants",
      "query",
      clientId,
      policyYearId,
      page.offset,
      page.limit,
      query,
    ],
    queryFn: () =>
      api.post<DependantList>(
        `/policy-years/${policyYearId}/dependant-query/list`,
        { query, ...page },
      ),
    enabled: !!policyYearId,
    placeholderData: (prev) => prev,
  });
}
